sys/aarch64/board.c

```diff
@@ -8,66 +8,109 @@
-    } else if (tag == ATAG_INITRD2) {
-      snprintf(buf, sizeof(buf), "rd_start=%d", atag->tag_initrd.start);
-      tokens = cmdline_extract_tokens(stk, buf, tokens);
-      snprintf(buf, sizeof(buf), "rd_size=%d", atag->tag_initrd.size);
-      tokens = cmdline_extract_tokens(stk, buf, tokens);
-    } else if (tag == ATAG_CMDLINE) {
-      tokens = cmdline_extract_tokens(stk, atag->tag_cmd.command, tokens);
-    }
-  }
+  return fdt32_to_cpu(prop[1]);
+}
+
+static int process_dtb_rd_start(void *dtb) {
+  int offset = dtb_offset(dtb, "/chosen");
+  int len;
+  const uint32_t *prop = fdt_getprop(dtb, offset, "linux,initrd-start", &len);
+  if (prop == NULL || (size_t)len < sizeof(uint32_t))
+    panic("Failed to get initrd-start from dtb!");
+
+  return fdt32_to_cpu(*prop);
+}
+
+static int process_dtb_rd_size(void *dtb) {
+  int offset = dtb_offset(dtb, "/chosen");
+  int len;
+  const uint32_t *prop = fdt_getprop(dtb, offset, "linux,initrd-end", &len);
+  if (prop == NULL || (size_t)len < sizeof(uint32_t))
+    panic("Failed to get initrd-start from dtb!");
+
+  return fdt32_to_cpu(*prop) - process_dtb_rd_start(dtb);
+}
+
+static const char *process_dtb_cmdline(void *dtb) {
+  int offset = dtb_offset(dtb, "/chosen");
+  const char *prop = fdt_getprop(dtb, offset, "bootargs", NULL);
+  if (prop == NULL)
+    panic("Failed to get cmdline from dtb!");
 
+  return prop;
+}
+
+static void process_dtb(char **tokens, kstack_t *stk, void *dtb) {
+  if (fdt_check_header(dtb))
+    panic("dtb incorrect header!");
+
+  char buf[32];
+  snprintf(buf, sizeof(buf), "memsize=%d", process_dtb_memsize(dtb));
+  tokens = cmdline_extract_tokens(stk, buf, tokens);
+  snprintf(buf, sizeof(buf), "rd_start=%d", process_dtb_rd_start(dtb));
+  tokens = cmdline_extract_tokens(stk, buf, tokens);
+  snprintf(buf, sizeof(buf), "rd_size=%d", process_dtb_rd_size(dtb));
+  tokens = cmdline_extract_tokens(stk, buf, tokens);
+  tokens = cmdline_extract_tokens(stk, process_dtb_cmdline(dtb), tokens);
   *tokens = NULL;
 }
 
-void *board_stack(void) {
+void *board_stack(paddr_t dtb) {
+  dtb_early_init(dtb, fdt_totalsize(PHYS_TO_DMAP(dtb)));
+
   kstack_t *stk = &thread0.td_kstack;
 
   thread0.td_uctx = kstack_alloc_s(stk, mcontext_t);
 
-  int ntokens = count_atags();
-  char **kenvp = kstack_alloc(stk, (ntokens + 2) * sizeof(char *));
-  process_atags(kenvp, stk);
+  /*
+   * NOTE: memsize, rd_start, rd_size, cmdline + 2 = 6
+   */
+  char **kenvp = kstack_alloc(stk, 6 * sizeof(char *));
+  process_dtb(kenvp, stk, (void *)PHYS_TO_DMAP(dtb));
   kstack_fix_bottom(stk);
   init_kenv(kenvp);
 
   return stk->stk_ptr;
 }
 
+typedef struct {
+  paddr_t start;
+  paddr_t end;
+} addr_range_t;
+
+static int addr_range_cmp(const void *_lhs, const void *_rhs) {
+  const addr_range_t *lhs = _lhs;
+  const addr_range_t *rhs = _rhs;
+
+  assert(lhs->start != rhs->start);
+
+  if (lhs->start < rhs->start)
+    return -1;
+  return 1;
+}
```

Comment:
Do we want an assert for overlapping memory addr?

Reply:
It is already asserted in physmem.